keps/sig-network/3786-migrate-shared-kube-proxy-code-to-staging/README.md

```diff
@@ -0,0 +1,301 @@
+to be incorporated back in tree, it became clear we needed to decompose this large task into smaller pieces. Therefore, we've decided to keep things simple and start by moving the existing shared kube-proxy code into staging where it can be iterated on and augmented to in an safe, consumable and incremental manner.
+
+### Goals
+
+- Move the [shared kube-proxy code k/k/pkg/proxy](https://github.com/kubernetes/kubernetes/tree/master/pkg/proxy), and relevant
+networking utilities (i.e `pkg/util/iptables`) into the kube-proxy [staging repository](https://github.com/kubernetes/kube-proxy).
+- Ensure all existing kube-proxy unit and e2e coverage remains the same or is improved.
+- Ensure the shared code can be easily vendored by external users to help write new out of tree service proxies.
+- Write documentation detailing how external consumers can utilize the kube-proxy staging library.
+
+### Non-Goals
+
+- Write any more new "in-tree" service proxies.
+- Make any incompatible architectural or deployment changes to the existing kube-proxy implementations.
+- Tackle any complex new deployment scenarios (This is solved by [KPNG](https://github.com/kubernetes-sigs/kpng/))
+
+## Proposal
+
+We propose to build a new library in the [kube-proxy staging repository](https://github.com/kubernetes/kube-proxy). This repository will be vendored by the core implementations and developers who want to build new service proxy implementations, providing them with:
+
+- A vendorable golang library that defines a few interfaces which can be easily implemented by a new service proxy, that responds to EndpointSlice and Service changes.
+- Documentation on how to build a kube proxy with the library, based on [So You Want To Write A Service Proxy...](https://github.com/kubernetes-sigs/kpng/blob/master/doc/service-proxy.md) and other similar resources.
+
+Not only will this make writing new backends easier, but through incremental changes and optimizations to the new library we hope to also improve the existing proxies, making [legacy bugs](https://github.com/kubernetes/kubernetes/issues/112604) easier to fix in the future.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a networking technology startup I want to easily make a new service proxy implementation without maintaining the logic of talking to the APIServer, caching its data, or calculating an abbreviated/proxy-focused representation of the Kubernetes networking state space.  I'd like a wholesale framework I can simply plug my custom dataplane oriented logic into.
+
+#### Story 2
+
+As a service proxy maintainer, I don't want to have to understand the in-tree internals of a networking backend in order to simulate or write core updates to the logic of the kube-proxy locally.
+
+#### Story 3
+
+As a Kubernetes developer I want to make maintaining the shared proxy code easier, and allow for updates to that code to be completed in a more incremental and well tested way.
```

Comment:
maintaining code and versioning in kubernetes/kubernetes is not easier, just the contrary.

There is also a misbelief that the testing comes for free, and that is not true, each SIG is responsible for maintaining, adding, monitoring and fixing its own tests, and this require a constant and periodic work, it can not be done sporadilly.

All the sig-network testing is maintained only by myself during last years, I tried to onboard more people, doing presentations, mentoring people 1on1s, but nobody sticks and people comes and goes leaving half baked tests, features or dashboards.
I'd like to see a plan and evidence that there will be more people involved on this work to avoid leaving the project in an undesired state

Reply:
> All the sig-network testing is maintained only by myself during last years, I tried to onboard more people, doing presentations, mentoring people 1on1s, but nobody sticks and people comes and goes leaving half baked tests, features or dashboards.

+1000 First of all thank YOU for all of this work :)  I think everyone who is stable in the community is feeling some bit of burnout/ trouble with this (I know I am in the network-policy group) 

So initially nothing would change with this move except for the location of the code.  Unless I'm missing something....we wouldn't be providing any new versioning guarantees, and we'll run all of the existing tests against the same code (maybe this will require some new hacks?) 

However now at least these bits can be used by external consumers, which IMO would hopefully lead to more involvement. 

> I'd like to see a plan and evidence that there will be more people involved on this work to avoid leaving the project in an undesired state

I totally agree maybe we can work on a plan here? At the root of it IMO if we expose functionality previously only exposed to core implementations, it would mean more users, which _hopefully_ would also mean more folks involved.

Reply:
> t would mean more users, which hopefully would also mean more folks involved.

heh, I never seen that happen but I love to stand corrected here

> we expose functionality previously only exposed to core implementations

that is my main fair, so we are going to have more features demands for niche use cases or to fix bug for implementation X that is not compatible with implementation Y